runtime/doc/lua.txt

```diff
@@ -2306,10 +2307,13 @@ find({names}, {opts})                                          *vim.fs.find()*
         (table) The paths of all matching files or directories
 
 normalize({path})                                         *vim.fs.normalize()*
-    Normalize a path to a standard format. A tilde (~) character at the
-    beginning of the path is expanded to the user's home directory and any
-    backslash (\) characters are converted to forward slashes (/). Environment
-    variables are also expanded.
+    Normalize a path to a standard format.
+
+    A tilde (~) character at the beginning of the path is expanded to the
+    user's home directory and any backslash (\) characters are converted to
```

Comment:
Perhaps we should do these kind of "resolve" tasks in `vim.fs.realpath` 

For comparison, 

- python has
    - [abspath](https://docs.python.org/3/library/os.path.html#os.path.abspath)
    - [normpath](https://docs.python.org/3/library/os.path.html#os.path.normpath)
    - [realpath](https://docs.python.org/3/library/os.path.html#os.path.realpath)
- nodejs has
    - [resolve](https://nodejs.org/api/path.html#pathresolvepaths)
    - [normalize](https://nodejs.org/api/path.html#pathnormalizepath)

Reply:
Maybe we don't need `realpath` at all and just use `normalize` for that purpose. One benefit of `realpath` is that it returns `nil` if the file does not exist which is useful in some situations. We could have `normalize` return a second boolean parameter that covers this use-case.

Reply:
It's probably useful to have separation of "resolve" and "normalize". After all, one can always do normalize(resolve())--but not the inverse if we don't have both.

Reply:
It's not clear to me what the difference is between `resolve` and `normalize` in your suggestion.

> * python has ...
> * nodejs has ...

I think taking inspiration from other stdlibs is a good idea as that lets users rely on existing mental models. We just need to pick one. Some of `vim.fs` is already based on penlight, which has `abspath` and `normpath`.

`abspath` does not use the `realpath` system function: it simply prepends the cwd to the given file (so it's really just `fnamemodify(path, ':p')`). This is useful in some cases, and might be worth adding.

`normpath` (what we call `normalize`) does the job of resolving `./` and `../` segments as well as symlinks. But it does not make the path absolute. So again, we would not be able to use `fs_realpath` for this either.

Let me know what API you'd like to see.

Reply:
> It's not clear to me what the difference is between resolve and normalize in your suggestion.

@gpanders in both of my above comments, intention was that `normalize` only does stuff that doesn't read from the environment. whereas `resolve` resolves against the environment/filesystem/etc.